### DISCUSSION TITLE
Bcrypt Hash does not work. Getting Error message when try to login

### DISCUSSION THREAD

**Discussion by squadramunter** [wg-easy/wg-easy#1148] [Q&A]
**Describe the bug**
When I follow the readme of how to create a bcrypt hash it does not work when applied to the docker run config with -e flag. I use PASSWORD_HASH= and use the docker hash version with two dollar signs. When applying it and try to login I get a error message instead and can't login.

**To Reproduce**
1. Create bcrypt hash using the readme
2. Applying the hashed version for docker using 2 dollar signs.
3. Try to login on the Web UI
4. Get a error message instead of logging in.

**Expected behavior**
Not getting a error message and getting inside the web ui.

**Desktop (please complete the following information):**
 - OS: Windows 11
 - Browser Chrome
 - Version 126.0.6478.127 (Official Build) (64-bit)

**Comment by XiepengLi** [comment]
On the contrary, I set a password and logged in directly without any authentication process.

**Comment by memesalot** [comment]
> On the contrary, I set a password and logged in directly without any authentication process.

Same

**Comment by optroodt** [comment]
> On the contrary, I set a password and logged in directly without any authentication process.

Same. Running using `docker compose`, initially configured without a `PASSWORD_HASH`, now configured one but I get no password prompt when accessing the web UI. I did a `docker compose down/up` to restart the container, but to no avail.

**Comment by pheiduck** [comment]
> On the contrary, I set a password and logged in directly without any authentication process.

Guys, The Password env is deprecated in the nightly images it's recommened to move over to `PASSWORD_HASH` and if it does not even work with double $ sign. Use "" instead.

**Comment by pheiduck** [comment]
I haven't experienced any issues with hashed password.

**Comment by vlaadoo** [comment]
> Use "" instead.

Can you give an example where to put the ""?

**Comment by pheiduck** [comment]
> Can you give an example where to put the ""?

Sure, `"$2y$10$3.0sbYnMvz98mnoy8J0lL.WxlkBFRDA2fK5eGhifayecC/okKDUuu"`

**Reply by vlaadoo** [reply]
I tried as you mentioned in example, but I still get an "Unauthorized" error when I try to log in with password. Could the problem be that I'm running a direct command to the docker and not from the compose file?

**Reply by eandersons** [reply]
@vlaadoo, because when `$` is used within double quotes in command line, then `$` and a string following it is interpolated as a variable. In this example the variables would be `$2`, `$1` and `$3`.

See for yourself: `echo "$2y$10$3.0sbYnMvz98mnoy8J0lL.WxlkBFRDA2fK5eGhifayecC/okKDUuu"`, which outputs `y0.0sbYnMvz98mnoy8J0lL.WxlkBFRDA2fK5eGhifayecC/okKDUuu` (`$2`, `$1` and `$3` are replaced with an empty string).

**Reply by vlaadoo** [reply]
@eandersons I tried to write in docker compose as `$$2y$$10$$3.0sbYnM...` and as `"$2y$10$3.0sbYnM..."` and also as `"$$2y$$10$$3.0sbYnM..."`.
But I still got an Unauthorized error each time.

**Reply by resulshm** [reply]
Same problem, have you solved this @vlaadoo ?

**Reply by vlaadoo** [reply]
@resulshm unfortunately not :(

**Reply by Alkalifax** [reply]
Ya lo resolvi, copien tal cual el output del comando 
docker run --rm -it ghcr.io/wg-easy/wg-easy wgpw
PASSWORD_HASH='$2a$12$HashPassword'

No usen el doble $ ni " usen comillas simples '

**Comment by FenixVoltres** [comment]
I just learned PASSWORD_HASH works only on :nightly builds. The python script provided with bcrypt also generates weird output for me, as AFAIK "hashed password" and "hashed password for an docker env" should only differ in a way that the latter has "$" chars duplicated, my in my case they are completely different. What I did and worked for me is to copy the former and manually replace "$" with "$$" when passing it to DOCKER env.

**Comment by squadramunter** [comment]
Maybe a suggestion to make a default hashed password and ask for new password after logging in? Would make it better for people struggling about this. WG-EASY is created for people that are not very experienced with building Wireguard configs themself. Would be nice to include such way to ask for password change on first login. Like Nginx Proxy Manager does as example.

**Comment by squadramunter** [comment]
> > Can you give an example where to put the ""?
> 
> Sure, `"$2y$10$3.0sbYnMvz98mnoy8J0lL.WxlkBFRDA2fK5eGhifayecC/okKDUuu"`

How does your Docker Compose file look like?

Can you descripe the process step by step how you did this? Looks like the manual for creating hashed passwords is missing something.

**Comment by lucasggit** [comment]
Solution for me :
when you use docker-compose format, you have to double ALL the "$" in the hash password :
docker run --rm ghcr.io/wg-easy/wg-easy wgpw MacronExplosion
PASSWORD_HASH='$2a$12$ktKot/61huQHPsBxMOc3S.pOptnWmqEynbM6Beps7998ZNPVgUNrS'

in my docker-compose :
- PASSWORD_HASH=$$2a$$12$$ktKot/61huQHPsBxMOc3S.pOptnWmqEynbM6Beps7998ZNPVgUNrS

**Reply by tabakovsasha** [reply]
Thank you, @lucasggit ! It's also helped me
